sdk/storage/azure-storage-queues/src/queue_client.cpp

```diff
@@ -0,0 +1,341 @@
+    m_pipeline = std::make_shared<Azure::Core::Http::_internal::HttpPipeline>(
+        newOptions,
+        _internal::QueueServicePackageName,
+        _detail::PackageVersion::ToString(),
+        std::move(perRetryPolicies),
+        std::move(perOperationPolicies));
+  }
+
+  QueueClient::QueueClient(
+      const std::string& queueUrl,
+      std::shared_ptr<Core::Credentials::TokenCredential> credential,
+      const QueueClientOptions& options)
+      : QueueClient(queueUrl, options)
+  {
+    std::vector<std::unique_ptr<Azure::Core::Http::Policies::HttpPolicy>> perRetryPolicies;
+    std::vector<std::unique_ptr<Azure::Core::Http::Policies::HttpPolicy>> perOperationPolicies;
+    perRetryPolicies.emplace_back(std::make_unique<_internal::StorageSwitchToSecondaryPolicy>(
+        m_queueUrl.GetHost(), options.SecondaryHostForRetryReads));
+    perRetryPolicies.emplace_back(std::make_unique<_internal::StoragePerRetryPolicy>());
+    {
+      Azure::Core::Credentials::TokenRequestContext tokenContext;
+      tokenContext.Scopes.emplace_back(_internal::StorageScope);
+      perRetryPolicies.emplace_back(
+          std::make_unique<Azure::Core::Http::Policies::_internal::BearerTokenAuthenticationPolicy>(
+              credential, tokenContext));
+    }
+    perOperationPolicies.emplace_back(
+        std::make_unique<_internal::StorageServiceVersionPolicy>(options.ApiVersion));
+    m_pipeline = std::make_shared<Azure::Core::Http::_internal::HttpPipeline>(
+        options,
+        _internal::QueueServicePackageName,
+        _detail::PackageVersion::ToString(),
+        std::move(perRetryPolicies),
+        std::move(perOperationPolicies));
+  }
+
+  QueueClient::QueueClient(const std::string& queueUrl, const QueueClientOptions& options)
+      : m_queueUrl(queueUrl)
+  {
+    std::vector<std::unique_ptr<Azure::Core::Http::Policies::HttpPolicy>> perRetryPolicies;
+    std::vector<std::unique_ptr<Azure::Core::Http::Policies::HttpPolicy>> perOperationPolicies;
+    perRetryPolicies.emplace_back(std::make_unique<_internal::StorageSwitchToSecondaryPolicy>(
+        m_queueUrl.GetHost(), options.SecondaryHostForRetryReads));
+    perRetryPolicies.emplace_back(std::make_unique<_internal::StoragePerRetryPolicy>());
+    perOperationPolicies.emplace_back(
+        std::make_unique<_internal::StorageServiceVersionPolicy>(options.ApiVersion));
+    m_pipeline = std::make_shared<Azure::Core::Http::_internal::HttpPipeline>(
+        options,
+        _internal::QueueServicePackageName,
+        _detail::PackageVersion::ToString(),
+        std::move(perRetryPolicies),
+        std::move(perOperationPolicies));
+  }
+
+  Azure::Response<Models::CreateQueueResult> QueueClient::Create(
+      const CreateQueueOptions& options,
+      const Azure::Core::Context& context) const
+  {
+    _detail::QueueRestClient::Queue::CreateQueueOptions protocolLayerOptions;
+    protocolLayerOptions.Metadata = options.Metadata;
```

Comment:
Is there a reason why we need a protocol-model-options and a public-model-options?  
I would expect the protocol layer to always use the public-model-options.

Reply:
Convenience layer API model and protocol layer API model was discussed like more than a year ago, and we landed on current model.

Reply:
Check the email titled _About the design on protocol layer_ from Jeffrey on Apr. 20, 2020.

Bring this up in next week's working group meeting if you want to overturn the conclusion.

Reply:
The rationale behind this design is that it requires human intervention to decide which parameters are mandatory and which are optional. But the input of auto-gen (which is usually swagger) doesn't contain this information. So there's no way auto-gen can know which parameters should be in parameter list and which should be in options bag.

Reply:
Since the protocol layer is implementation detail, I would make all methods to take only const& to all values directly, without introducing new models. That would save the extra-copy.  And the auto-gen would just flatten the models into a full list of input parameters for the method.
Anyway, This is not something I would ask to be changed at this point. It's late for that. I can live with this. Thank you for explaining it.